File: backend/src/api/routes/accounts.py
"""Account planning API routes for US-941.

Endpoints:
- GET  /accounts              — List accounts (territory view)
- GET  /accounts/{id}/plan    — Get or generate account plan
- PUT  /accounts/{id}/plan    — Update account plan strategy
- GET  /accounts/territory    — Territory overview (alias with stats)
- GET  /accounts/forecast     — Pipeline forecast
- GET  /accounts/quota        — Get quota records
- POST /accounts/quota        — Set/update quota
"""

import logging
from typing import Any

from fastapi import APIRouter, HTTPException, Query

from src.api.deps import CurrentUser
from src.models.account_planning import AccountPlanUpdate, QuotaSet
from src.services.account_planning_service import AccountPlanningService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/accounts", tags=["accounts"])


def _get_service() -> AccountPlanningService:
    """Get account planning service instance."""
    return AccountPlanningService()


# --- Static routes MUST come before parametric routes ---


@router.get("/territory")
async def get_territory(
    current_user: CurrentUser,
    stage: str | None = Query(None, description="Filter by lifecycle stage"),
    sort_by: str = Query("last_activity_at", description="Sort column"),
    limit: int = Query(50, ge=1, le=200, description="Max accounts"),
) -> dict[str, Any]:
    """Get territory overview with account list and summary stats.

    Returns accounts plus aggregate stats for the territory dashboard header.
    """
    service = _get_service()
    accounts = await service.list_accounts(current_user.id, stage, sort_by, limit)

    total_value = sum(float(a.get("expected_value") or 0) for a in accounts)
    avg_health = (
        round(sum(a.get("health_score", 0) for a in accounts) / len(accounts)) if accounts else 0
    )
    stage_counts: dict[str, int] = {}
    for a in accounts:
        s = a.get("lifecycle_stage", "lead")
        stage_counts[s] = stage_counts.get(s, 0) + 1

    logger.info(
        "Territory overview retrieved",
        extra={"user_id": current_user.id, "account_count": len(accounts)},
    )

    return {
        "accounts": accounts,
        "stats": {
            "total_accounts": len(accounts),
            "total_value": total_value,
            "avg_health": avg_health,
            "stage_counts": stage_counts,
        },
    }


@router.get("/forecast")
async def get_forecast(current_user: CurrentUser) -> dict[str, Any]:
    """Get pipeline forecast based on lead health scores and expected values."""
    service = _get_service()
    return await service.get_forecast(current_user.id)


@router.get("/quota")
async def get_quota(
    current_user: CurrentUser,
    period: str | None = Query(None, description="Filter by period"),
) -> list[dict[str, Any]]:
    """Get quota tracking records."""
    service = _get_service()
    return await service.get_quota(current_user.id, period)


@router.post("/quota")
async def set_quota(
    data: QuotaSet,
    current_user: CurrentUser,
) -> dict[str, Any]:
    """Set or update a quota target for a period."""
    service = _get_service()
    quota = await service.set_quota(current_user.id, data.period, data.target_value)
    logger.info(
        "Quota set via API",
        extra={"user_id": current_user.id, "period": data.period},
    )
    return quota


# --- Parametric routes ---


@router.get("")
async def list_accounts(
    current_user: CurrentUser,
    stage: str | None = Query(None, description="Filter by lifecycle stage"),
    sort_by: str = Query("last_activity_at", description="Sort column"),
    limit: int = Query(50, ge=1, le=200, description="Max accounts"),
) -> list[dict[str, Any]]:
    """List accounts from Lead Memory."""
    service = _get_service()
    return await service.list_accounts(current_user.id, stage, sort_by, limit)


@router.get("/{lead_id}/plan")
async def get_account_plan(
    lead_id: str,
    current_user: CurrentUser,
) -> dict[str, Any]:
    """Get or auto-generate an account plan for a lead."""
    service = _get_service()
    plan = await service.get_or_generate_plan(current_user.id, lead_id)
    if plan is None:
        raise HTTPException(status_code=404, detail="Account not found")
    return plan


@router.put("/{lead_id}/plan")
async def update_account_plan(
    lead_id: str,
    data: AccountPlanUpdate,
    current_user: CurrentUser,
) -> dict[str, Any]:
    """Update account plan strategy text (user edits)."""
    service = _get_service()
    plan = await service.update_plan(current_user.id, lead_id, data.strategy)
    if plan is None:
        raise HTTPException(status_code=404, detail="Account plan not found")
    logger.info(
        "Account plan updated via API",
        extra={"user_id": current_user.id, "lead_id": lead_id},
    )
    return plan

File: frontend/src/api/accounts.ts
import { apiClient } from "./client";

// --- Types ---

export interface AccountListItem {
  id: string;
  company_name: string;
  lifecycle_stage: string;
  status: string;
  health_score: number;
  expected_value: number | null;
  last_activity_at: string | null;
  tags: string[];
  next_action: string | null;
}

export interface TerritoryStats {
  total_accounts: number;
  total_value: number;
  avg_health: number;
  stage_counts: Record<string, number>;
}

export interface TerritoryResponse {
  accounts: AccountListItem[];
  stats: TerritoryStats;
}

export interface AccountPlan {
  id: string;
  user_id: string;
  lead_memory_id: string;
  strategy: string;
  next_actions: Array<{
    action: string;
    priority: "high" | "medium" | "low";
    due_in_days?: number;
  }>;
  stakeholder_summary: {
    champion?: string | null;
    decision_maker?: string | null;
    key_risk?: string;
  };
  generated_at: string;
  updated_at: string;
}

export interface ForecastStage {
  stage: string;
  count: number;
  total_value: number;
  weighted_value: number;
}

export interface ForecastResponse {
  stages: ForecastStage[];
  total_pipeline: number;
  weighted_pipeline: number;
}

export interface Quota {
  id: string;
  user_id: string;
  period: string;
  target_value: number;
  actual_value: number;
  created_at: string;
  updated_at: string;
}

// --- API functions ---

export async function listAccounts(
  stage?: string,
  sortBy?: string,
  limit?: number
): Promise<AccountListItem[]> {
  const params = new URLSearchParams();
  if (stage) params.append("stage", stage);
  if (sortBy) params.append("sort_by", sortBy);
  if (limit) params.append("limit", limit.toString());
  const url = params.toString() ? `/accounts?${params}` : "/accounts";
  const response = await apiClient.get<AccountListItem[]>(url);
  return response.data;
}

export async function getTerritory(
  stage?: string,
  sortBy?: string,
  limit?: number
): Promise<TerritoryResponse> {
  const params = new URLSearchParams();
  if (stage) params.append("stage", stage);
  if (sortBy) params.append("sort_by", sortBy);
  if (limit) params.append("limit", limit.toString());
  const url = params.toString()
    ? `/accounts/territory?${params}`
    : "/accounts/territory";
  const response = await apiClient.get<TerritoryResponse>(url);
  return response.data;
}

export async function getAccountPlan(leadId: string): Promise<AccountPlan> {
  const response = await apiClient.get<AccountPlan>(
    `/accounts/${leadId}/plan`
  );
  return response.data;
}

export async function updateAccountPlan(
  leadId: string,
  strategy: string
): Promise<AccountPlan> {
  const response = await apiClient.put<AccountPlan>(
    `/accounts/${leadId}/plan`,
    { strategy }
  );
  return response.data;
}

export async function getForecast(): Promise<ForecastResponse> {
  const response = await apiClient.get<ForecastResponse>(
    "/accounts/forecast"
  );
  return response.data;
}

export async function getQuotas(period?: string): Promise<Quota[]> {
  const params = period ? `?period=${period}` : "";
  const response = await apiClient.get<Quota[]>(
    `/accounts/quota${params}`
  );
  return response.data;
}

export async function setQuota(
  period: string,
  targetValue: number
): Promise<Quota> {
  const response = await apiClient.post<Quota>("/accounts/quota", {
    period,
    target_value: targetValue,
  });
  return response.data;
}
